File: chatuikit/src/main/java/com/cometchat/chatuikit/shared/resources/utils/AudioPlayer.java
package com.cometchat.chatuikit.shared.resources.utils;

import android.content.Context;
import android.media.MediaPlayer;

import com.cometchat.chatuikit.logger.CometChatLogger;

public class AudioPlayer {
    private static final String TAG = AudioPlayer.class.getSimpleName();

    private static AudioPlayer instance;
    private final MediaPlayer mediaPlayer;
    private boolean isPrepared;
    private boolean isPaused;
    private MediaPlayer.OnCompletionListener completionListener;
    private Runnable onPlaybackStartListener;
    private Context context;

    private AudioPlayer() {
        mediaPlayer = new MediaPlayer();
    }

    public static AudioPlayer getInstance() {
        if (instance == null) {
            instance = new AudioPlayer();
        }
        return instance;
    }

    public void setAudioUrl(String url, MediaPlayer.OnPreparedListener preparedListener, MediaPlayer.OnCompletionListener completionListener) {
        try {
            reset();
            this.completionListener = completionListener;
            mediaPlayer.setDataSource(url);
            mediaPlayer.setOnPreparedListener(mp -> {
                isPrepared = true;
                if (preparedListener != null) preparedListener.onPrepared(mediaPlayer);
            });
            mediaPlayer.setOnCompletionListener(mediaPlayer -> {
                if (completionListener != null) completionListener.onCompletion(mediaPlayer);
            });
            mediaPlayer.prepareAsync();
        } catch (Exception e) {
            CometChatLogger.e(TAG, e.toString());
        }
    }

    public void reset() {
        mediaPlayer.reset();
        if (completionListener != null)
            completionListener.onCompletion(mediaPlayer);
        isPrepared = false;
        isPaused = false;
    }

    public void start() {
        if (!mediaPlayer.isPlaying() && isPrepared) {
            mediaPlayer.start();
            if (onPlaybackStartListener != null) {
                onPlaybackStartListener.run();
            }
        }
    }

    public void stop() {
        if (mediaPlayer.isPlaying()) {
            mediaPlayer.stop();
            isPrepared = false;
            isPaused = false;
        }
    }

    public void pause() {
        if (mediaPlayer.isPlaying() && isPrepared) {
            mediaPlayer.pause();
            isPaused = true;
        }
    }

    public void resume() {
        if (isPaused && isPrepared) {
            mediaPlayer.start();
            isPaused = false;
            if (onPlaybackStartListener != null) {
                onPlaybackStartListener.run();
            }
        }
    }

    public boolean isPaused() {
        return isPaused;
    }

    public boolean isPlaying() {
        return mediaPlayer.isPlaying();
    }

    public MediaPlayer getMediaPlayer() {
        return mediaPlayer;
    }

    /**
     * Sets a listener to be notified when playback starts.
     * Used by inline audio recorder to stop its own playback.
     */
    public void setOnPlaybackStartListener(Runnable listener) {
        this.onPlaybackStartListener = listener;
    }

    /**
     * Sets the context for MediaPlayer operations.
     */
    public void setContext(Context context) {
        this.context = context;
    }

    /**
     * Plays audio from a local file path.
     * Used when the file is already downloaded (e.g., from local path in metadata).
     */
    public void playFromLocalFile(String filePath, MediaPlayer.OnPreparedListener preparedListener, MediaPlayer.OnCompletionListener completionListener) {
        try {
            reset();
            this.completionListener = completionListener;
            mediaPlayer.setDataSource(filePath);
            mediaPlayer.setOnPreparedListener(mp -> {
                isPrepared = true;
                if (preparedListener != null) preparedListener.onPrepared(mediaPlayer);
            });
            mediaPlayer.setOnCompletionListener(mp -> {
                if (completionListener != null) completionListener.onCompletion(mp);
            });
            mediaPlayer.prepareAsync();
        } catch (Exception e) {
            CometChatLogger.e(TAG, "Error playing from local file: " + e.toString());
        }
    }
}

File: chatuikit-core/src/main/java/com/cometchat/uikit/core/utils/AudioPlayer.kt
package com.cometchat.uikit.core.utils

import android.media.MediaPlayer
import android.util.Log

/**
 * Singleton audio player for managing audio playback across the UI Kit.
 *
 * Ensures only one audio plays at a time — when a new audio is started,
 * the previous one is automatically stopped and its completion listener is fired.
 *
 * Shared across chatuikit-kotlin and chatuikit-jetpack modules.
 */
class AudioPlayer private constructor() {

    private val mediaPlayer = MediaPlayer()
    private var isPrepared = false
    private var completionListener: MediaPlayer.OnCompletionListener? = null

    companion object {
        private const val TAG = "AudioPlayer"

        @Volatile
        private var instance: AudioPlayer? = null

        @JvmStatic
        fun getInstance(): AudioPlayer {
            return instance ?: synchronized(this) {
                instance ?: AudioPlayer().also { instance = it }
            }
        }
    }

    /**
     * Sets the audio URL and prepares the player.
     * Automatically resets any currently playing audio first.
     *
     * @param url The URL of the audio file
     * @param preparedListener Callback when the player is prepared
     * @param completionListener Callback when playback completes
     */
    fun setAudioUrl(
        url: String,
        preparedListener: MediaPlayer.OnPreparedListener?,
        completionListener: MediaPlayer.OnCompletionListener?
    ) {
        try {
            reset()
            mediaPlayer.setDataSource(url)
            mediaPlayer.prepare()
            this.completionListener = completionListener
            mediaPlayer.setOnPreparedListener { mp ->
                isPrepared = true
                mediaPlayer.start()
                preparedListener?.onPrepared(mp)
            }
            mediaPlayer.setOnCompletionListener { mp ->
                completionListener?.onCompletion(mp)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error setting audio URL: ${e.message}")
        }
    }

    /**
     * Resets the media player.
     * Fires the completion listener of the previously playing audio.
     */
    fun reset() {
        mediaPlayer.reset()
        completionListener?.onCompletion(mediaPlayer)
        isPrepared = false
    }

    /**
     * Starts playback if prepared and not already playing.
     */
    fun start() {
        if (!mediaPlayer.isPlaying && isPrepared) {
            mediaPlayer.start()
        }
    }

    /**
     * Stops playback.
     */
    fun stop() {
        if (mediaPlayer.isPlaying) {
            mediaPlayer.stop()
            isPrepared = false
        }
    }

    /**
     * Checks if audio is currently playing.
     */
    fun isPlaying(): Boolean = mediaPlayer.isPlaying

    /**
     * Gets the underlying MediaPlayer instance.
     */
    fun getMediaPlayer(): MediaPlayer = mediaPlayer
}
